AX Web Inspector Panel: Switch form controls report "Checked: true/false" rather than something like "State: on/off"

twilco opened the pull request:
#### 132e20aa121367262caab344e193535bef849955
<pre>
AX Web Inspector Panel: Switch form controls report &quot;Checked: true/false&quot; rather than something like &quot;State: on/off&quot;
<a href="https://bugs.webkit.org/show_bug.cgi?id=274846">https://bugs.webkit.org/show_bug.cgi?id=274846</a>
<a href="https://rdar.apple.com/128952449">rdar://128952449</a>

Reviewed by Devin Rousso.

Switches are like checkboxes without a mixed state, and are exposed in the web platform in this way — HTML switches
are created via `input type=&quot;checkbox&quot;` plus the `switch` attribute, and are &quot;enabled&quot; via the `checked` attribute.
And ARIA switches (`role=&quot;switch&quot;`) use `aria-checked`.

A side effect of this is that when we added support for switch controls, we forgot to update Web Inspector to handle
the nuanced difference between these types of controls, and thus switch controls report a &quot;Checked: true/false&quot; state
rather than &quot;State: on/off&quot;, which is more appropriate terminology. This patch addresses this issue.

* LayoutTests/inspector/dom/getAccessibilityPropertiesForNode-expected.txt:
* LayoutTests/inspector/dom/getAccessibilityPropertiesForNode.html:
* Source/JavaScriptCore/inspector/protocol/DOM.json:
* Source/WebCore/inspector/agents/InspectorDOMAgent.cpp:
(WebCore::InspectorDOMAgent::buildObjectForAccessibilityProperties):
* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:
* Source/WebInspectorUI/UserInterface/Models/DOMNode.js:
(WI.DOMNode.prototype.accessibilityProperties.accessibilityPropertiesCallback):
(WI.DOMNode.prototype.accessibilityProperties):
* Source/WebInspectorUI/UserInterface/Views/DOMNodeDetailsSidebarPanel.js:
(WI.DOMNodeDetailsSidebarPanel.prototype.initialLayout):
(WI.DOMNodeDetailsSidebarPanel.prototype._refreshAccessibility.accessibilityPropertiesCallback):
(WI.DOMNodeDetailsSidebarPanel.prototype._refreshAccessibility):

Canonical link: <a href="https://commits.webkit.org/279772@main">https://commits.webkit.org/279772@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bc6a4496e8d71868ab10cd4b089afad339593b9c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/54228 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/33617 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/6770 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/57505 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/4954 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/56531 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/41145 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/4901 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/43928 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/3322 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/56324 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/31876 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/46964 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/25070 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/28688 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/4291 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/3101 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/47608 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/50558 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/4497 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/59097 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/53736 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/29440 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/4653 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/51347 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/30606 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/47072 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/50710 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11907 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/31581 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/66037 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/30389 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/12580 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

dcrousso commented on Source/WebInspectorUI/UserInterface/Models/DOMNode.js:
```diff
@@ -935,7 +935,8 @@ WI.DOMNode = class DOMNode extends WI.Object
                     selected: accessibilityProperties.selected,
-                    selectedChildNodeIds: accessibilityProperties.selectedChildNodeIds
+                    selectedChildNodeIds: accessibilityProperties.selectedChildNodeIds,
+                    switchState: accessibilityProperties.switchState
```
Style: please add a trailing `,` to this line

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/DOMNodeDetailsSidebarPanel.js:
```diff
@@ -546,6 +547,14 @@ WI.DOMNodeDetailsSidebarPanel = class DOMNodeDetailsSidebarPanel extends WI.DOMD
                         checked = WI.UIString("No");
                 }
 
+                var switchState = "";
```
Style: please use `let`

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/DOMNodeDetailsSidebarPanel.js:
```diff
@@ -728,6 +737,7 @@ WI.DOMNodeDetailsSidebarPanel = class DOMNodeDetailsSidebarPanel extends WI.DOMD
                 this._accessibilityNodeInvalidRow.value = invalid;
                 this._accessibilityNodeLabelRow.value = label;
                 this._accessibilityNodeLiveRegionStatusRow.value = liveRegionStatusNode || liveRegionStatus;
+                this._accessibilityNodeSwitchStateRow.value = switchState;
```
this will be hidden when it has no value (i.e. `""`) right?  if not, we should make sure this is only shown when it has a value

twilco replied:
Correct, that's the intention! It behaves this way with the PR as-is (the "State" row shows for switch controls, and no other type of element, as they have `switchState === ""`), but please let me know if there's something else that needs to be done or I've otherwise missed something.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/DOMNodeDetailsSidebarPanel.js:
```diff
@@ -546,6 +547,14 @@ WI.DOMNodeDetailsSidebarPanel = class DOMNodeDetailsSidebarPanel extends WI.DOMD
+                        switchState = WI.UIString("On");
+                    else
+                        switchState = WI.UIString("Off");
+                }
```
rather than explicitly checking for `undefined` it'd be much simpler to do this
```js
// COMPATIBILITY (iOS X.0): DOM.AccessibilityProperties.switchState did not exist yet.
if (InspectorBackend.Enum.DOM.AccessibilityPropertiesSwitchState) {
    switch (accessibilityProperties.switchState) {
    case InspectorBackend.Enum.DOM.AccessibilityPropertiesSwitchState.On:
        switchState = WI.UIString("On");
        break;
    case InspectorBackend.Enum.DOM.AccessibilityPropertiesSwitchState.Off:
        switchState = WI.UIString("Off");
        break;
    }
}
```

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/DOMNodeDetailsSidebarPanel.js:
```diff
@@ -777,7 +787,8 @@ WI.DOMNodeDetailsSidebarPanel = class DOMNodeDetailsSidebarPanel extends WI.DOMD
                     this._accessibilityNodeReadonlyRow,
-                    this._accessibilityNodeSelectedRow
+                    this._accessibilityNodeSelectedRow,
+                    this._accessibilityNodeSwitchStateRow
```
Style: please add a trailing , to this line

rcaliman-apple commented on Source/WebCore/inspector/agents/InspectorDOMAgent.cpp:
```diff
@@ -2423,8 +2434,14 @@ Ref<Inspector::Protocol::DOM::AccessibilityProperties> InspectorDOMAgent::buildO
-        if (supportsChecked)
+
+        // Switches `supportsChecked` (the underlying implementation is mostly shared), but should report
+        // a switch state and not a checked state.
```
Perhaps slightly clearer:
```diff
- the underlying implementation is mostly shared
+ the underlying implementation is mostly shared with checkboxes 
```

rcaliman-apple commented on Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:
```diff
@@ -1190,6 +1190,7 @@ localizedStrings["Observer Callback"] = "Observer Callback";
 localizedStrings["Observer Handlers:"] = "Observer Handlers:";
 localizedStrings["Observers:"] = "Observers:";
 localizedStrings["Off"] = "Off";
+localizedStrings["On"] = "On";
```
NIT: Can you please add a comment to help translators identify what "On" represents?

The translation can vary according to context.
Thinking in Romanian here, the former is more common, but the later is more suitable for a switch:

On = Pornit | Activ
Off = Oprit | Inactiv

I'm sure that's the case for other languages too.

twilco replied:
Good point, thanks! Fixed now.

rcaliman-apple reviewed (commented):
This looks solid. Thank you @twilco for bringing support in Web Inspector for switch inputs!

Still needs an official reviewer to approve (@dcrousso or @patrickangle)

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/DOMNodeDetailsSidebarPanel.js:
```diff
@@ -546,6 +547,19 @@ WI.DOMNodeDetailsSidebarPanel = class DOMNodeDetailsSidebarPanel extends WI.DOMD
+                if (InspectorBackend.Enum.DOM.AccessibilityPropertiesSwitchState) {
+                    switch (accessibilityProperties.switchState) {
+                    case InspectorBackend.Enum.DOM.AccessibilityPropertiesSwitchState.On:
+                        switchState = WI.UIString("Switch State @ On");
```
unfortunately this is not how `WI.UIString` works.  it's `WI.UIString(string, key, comment)`, so in this case you'd probably want something like
```js
WI.UIString("On", "On @ Switch State", "Label indicating that a input of type switch is on.");
```

twilco replied:
Ahh, thank you for catching this!

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/DOMNodeDetailsSidebarPanel.js:
```diff
@@ -546,6 +547,19 @@ WI.DOMNodeDetailsSidebarPanel = class DOMNodeDetailsSidebarPanel extends WI.DOMD
+                        switchState = WI.UIString("Switch State @ On");
+                        break;
+                    case InspectorBackend.Enum.DOM.AccessibilityPropertiesSwitchState.Off:
+                        switchState = WI.UIString("Switch State @ Off");
```
ditto
```js
WI.UIString("Off", "Off @ Switch State", "Label indicating that a input of type switch is off.");
```

dcrousso commented on Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:
```diff
@@ -1190,6 +1190,10 @@ localizedStrings["Observer Callback"] = "Observer Callback";
+/* The "on" state of switch controls. */
+localizedStrings["On @ Switch State"] = "On";
+/* The "off" state of switch controls. */
+localizedStrings["Off @ Switch State"] = "Off";
```
this does not match what's actually in the code

FYI this file is intended to be autogenerated from running
```bash
./Tools/Scripts/extract-localizable-js-strings --utf8 ./Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js ./Source/WebInspectorUI/UserInterface
```

dcrousso reviewed (approved):
nice work!  :)

twilco committed:
AX Web Inspector Panel: Switch form controls report "Checked: true/false" rather than something like "State: on/off"
https://bugs.webkit.org/show_bug.cgi?id=274846
rdar://128952449

Reviewed by Devin Rousso.

Switches are like checkboxes without a mixed state, and are exposed in the web platform in this way — HTML switches
are created via `input type="checkbox"` plus the `switch` attribute, and are "enabled" via the `checked` attribute.
And ARIA switches (`role="switch"`) use `aria-checked`.

A side effect of this is that when we added support for switch controls, we forgot to update Web Inspector to handle
the nuanced difference between these types of controls, and thus switch controls report a "Checked: true/false" state
rather than "State: on/off", which is more appropriate terminology. This patch addresses this issue.

* LayoutTests/inspector/dom/getAccessibilityPropertiesForNode-expected.txt:
* LayoutTests/inspector/dom/getAccessibilityPropertiesForNode.html:
* Source/JavaScriptCore/inspector/protocol/DOM.json:
* Source/WebCore/inspector/agents/InspectorDOMAgent.cpp:
(WebCore::InspectorDOMAgent::buildObjectForAccessibilityProperties):
* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:
* Source/WebInspectorUI/UserInterface/Models/DOMNode.js:
(WI.DOMNode.prototype.accessibilityProperties.accessibilityPropertiesCallback):
(WI.DOMNode.prototype.accessibilityProperties):
* Source/WebInspectorUI/UserInterface/Views/DOMNodeDetailsSidebarPanel.js:
(WI.DOMNodeDetailsSidebarPanel.prototype.initialLayout):
(WI.DOMNodeDetailsSidebarPanel.prototype._refreshAccessibility.accessibilityPropertiesCallback):
(WI.DOMNodeDetailsSidebarPanel.prototype._refreshAccessibility):

Canonical link: https://commits.webkit.org/279772@main
